/*!
* @file
* @brief Device
*
* @copyright Copyright (C) University of Bristol 2019
*    This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
*    If a copy of the MPL was not distributed with this file, You can obtain one at https://mozilla.org/MPL/2.0/.
*    See LICENSE file for details.
* @date 15/04/2019
* @author Richard Collins <richard.collins@bristol.ac.uk>
*/

syntax = "proto3";
import public "google/protobuf/empty.proto";

package cqp.remote;

message Pair {
    string name = 1;
    string value = 2;
}

/** Defines which side of a pair of devices
*/
message Side {
    /** Encapsulated type for scoping
    */
    enum Type {
        Alice = 0;
        Bob = 1;
        /// For device independent QKD
        Any = 2;
    }
    /// Value
    Type value = 1;
}

// possible parameters
// pulse width
// avg. photons per pulse (<1.0)
// filter bias of bases
// wavelength + tolerance
// photon state purity

/// values which describe the optical link
message DeviceConfig {
    /// arbritary parameters
    repeated Pair parameters = 1;
    /// The loss in dB for this link
    double lineAttenuation = 2;
    uint64 photonsPerBurst = 3;
    uint32 repitionRate = 4;

    /// The internal host name for use between devices
    string hostname = 5;
    /// Internal port number for use between wrappers
    int32 portNumber = 6;
    /// The kind of device the wrapper is connected to
    Side.Type side = 7;
};
